woocommerce/woocommerce-ios
# [Order List Redesign] Support starter view in SearchViewController

Refs #956. 

This is the first part of implementing this view:

<img src="https://user-images.githubusercontent.com/198826/73486280-0e0fe480-4362-11ea-9756-c9600647cf1b.png" width="240">

The view is shown when there is no keyword entered in the search bar. 

This PR is only for the `SearchViewController` architectural change to support the above UI.

## The State Machine

To support this, I added [2 additional `State` values](https://github.com/woocommerce/woocommerce-ios/blob/0dc49e9409e694af8b9cc5d74995013d91d7087e/WooCommerce/Classes/ViewRelated/Search/SearchViewController.swift#L489-L510) in `SearchViewController`:

- `notInitialized`: The default if `viewDidLoad` hasn't been called yet. 
- `starter`: The state if there is no search keyword and `SearchUICommand` **provides a starter `ViewController` via [`createStarterViewController()`](https://github.com/woocommerce/woocommerce-ios/blob/0dc49e9409e694af8b9cc5d74995013d91d7087e/WooCommerce/Classes/ViewRelated/Search/SearchUICommand.swift#L19-L29).

![image](https://user-images.githubusercontent.com/198826/73495642-2f79cc00-4374-11ea-9abe-0ee141dd509b.png)

## View Architecture

I expanded `SearchUICommand` to allow classes to optionally return a `UIViewController` as the "starter" `ViewController`:

https://github.com/woocommerce/woocommerce-ios/blob/0dc49e9409e694af8b9cc5d74995013d91d7087e/WooCommerce/Classes/ViewRelated/Search/SearchUICommand.swift#L19-L29

If this is non-`nil`, that `UIViewController` will be added to the `SearchViewController` as a **child `ViewController`*. 

![image](https://user-images.githubusercontent.com/198826/73489401-c12f0c80-4367-11ea-92de-948785aae5d0.png)

The `SearchViewController` will alternate between the starter `ViewController`'s `view` and the `.tableView` depending on the current `state`.

## Why not Reuse the Existing `UITableView`? 

I intentionally avoided trying to refactor `SearchViewController` so that it will show different cells depending on the `state`. I believe that making `SearchViewController` a **container `ViewController`** is a better design:

- More opportunities for encapsulation. The code to show the starter `ViewController` is encapsulated in a separate class and, potentially, testable. 
- As we experienced in https://github.com/wordpress-mobile/WordPress-iOS/pull/12974, switching `dataSource` and alternating between different `UITableViewCell` types can lead to complex race-condition crashes, even though everything is running on the main thread. 

## Unit Tests

I started [writing a unit test for `SearchViewController`](https://gist.github.com/shiki/a81c215bc6f9921d9889fedf514f07f3) until I realized that everything is `private`. I believe in order to this properly, we should move most of the logic (e.g. state transitions) to a `ViewModel` and test that instead. 

## Reviewing 

This PR does not include the order statuses view shown above. Rather, it only includes the code to support that view. Hence, please focus the review on the architectural decisions made. 🙂 

After reviewing this, please review #1805 next. 🙂 

## Testing 

1. **Use the branch in #1805** which implements [`OrderSearchStarterViewController`](https://github.com/woocommerce/woocommerce-ios/blob/3af0a89433c31336287a86325eb032577cb34553/WooCommerce/Classes/ViewRelated/Search/Order/OrderSearchStarterViewController.swift)
2. Navigate to Orders → Search
3. Confirm that the Order Statuses view is shown when there is no keyword entered.
4. Confirm that the Order Statuses view is hidden when there are search results. 

## Submitter Checklist
